pkg/policy/api/utils.go

```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+				if err != nil {
+					return err
+				}
+
+				if strings.ToLower(string(port.Protocol)) == strings.ToLower(string(portProtocol.Protocol)) && int(port.Port) == numericPort {
+					found = true
+					break loop
+				}
+			}
+		}
+		if !found {
+			portRule.Ports = append(portRule.Ports, PortProtocol{
+				Port:     strconv.Itoa(int(port.Port)),
+				Protocol: L4Proto(strings.ToUpper(string(port.Protocol))),
+			})
+		}
+	}
+
+	if len(portRule.Ports) > 0 {
+		egress.ToPorts = append(egress.ToPorts, portRule)
+	}
+
+	return nil
+}
+
+// DeleteGeneratedToServiceRulesFromEndpoint removes all ToService-based ToCIDR
+// rules generated from endpoint
+func (e *EgressRule) DeleteGeneratedToServiceRulesFromEndpoint(serviceInfo types.K8sServiceNamespace, endpoint types.K8sServiceEndpoint) error {
+	for _, service := range e.ToServices {
+		// TODO: match services by labels
+		if service.K8sService == serviceInfo {
+			if err := deleteToCidrFromEndpoint(e, endpoint); err != nil {
+				return err
+			}
+			// TODO: deleteToPortsFromEndpoint when ToPorts and ToCIDR are compatible
+		}
+	}
+	return nil
+}
+
+// deleteToCidrFromEndpoint takes an egress rule and removes ToCIDR rules matching endpoint
+func deleteToCidrFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
+	newToCIDR := make([]CIDR, 0, len(egress.ToCIDR))
+
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
+		for _, c := range egress.ToCIDR {
+			_, cidr, err := net.ParseCIDR(string(c))
+			if err != nil {
+				return err
+			}
+			if !cidr.Contains(epIP) {
+				//if endpoint is not in CIDR it's ok to retain it
+				newToCIDR = append(newToCIDR, c)
+			}
+		}
+	}
+
+	egress.ToCIDR = newToCIDR
+
+	return nil
+}
+
+// deleteToPortsFromEndpoint takes an egress rule and removes ToPorts rules matching endpoint
+func deleteToPortsFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
+	newPortRules := make([]PortRule, 0, len(egress.ToPorts))
+
+	for _, port := range endpoint.Ports {
+		for _, portRule := range egress.ToPorts {
+			for _, portProtocol := range portRule.Ports {
+				numericPort, err := strconv.Atoi(portProtocol.Port)
+				if err != nil {
+					return err
+				}
+
+				if !(strings.ToLower(string(port.Protocol)) == strings.ToLower(string(portProtocol.Protocol)) && int(port.Port) == numericPort) {
+					newPortRules = append(newPortRules, portRule)
+				}
+			}
+		}
+	}
+
+	egress.ToPorts = newPortRules
+
+	return nil
+}
+
+// GenerateEgressRulesFromEndpoints matches all egress rules against provided endpoint data and generates ToCIDR and ToPort rules
+func (r Rules) GenerateEgressRulesFromEndpoints(endpoints map[types.K8sServiceNamespace]*types.K8sServiceEndpoint) error {
+	for _, rule := range r {
+		for index := range rule.Egress {
+			for ns, ep := range endpoints {
+				if err := rule.Egress[index].GenerateToServiceRulesFromEndpoint(ns, *ep); err != nil {
+					return err
```

Comment:
In case of an error you might already changed some rules in the policy repo. Can you prevent this?